segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/IndexReader.java

```diff
@@ -0,0 +1,158 @@
+ *
+ * {@link TableBucket} Extended Attribute Implementation:
+ * - The Key Hash (a {@link UUID}) is used as a direct Extended Attribute entry. The {@link KeyHasher}, when generating
+ * this value, resolves any collisions with core attributes or Backpointer attributes.
+ * - The value of this attribute is the Segment Offset of the Table Entry with the highest offset of the {@link TableBucket}.
+ *
+ * Backpointers Extended Attribute Implementation:
+ * - Backpointers are used to resolve collisions for those Table Entries whose Keys have the same Key Hash.
+ * - Backpointers are links from the Offset of a Table Entry to the Offset of a previous Table Entry in the same {@link TableBucket}.
+ * -- They begin from the {@link TableBucket}'s SegmentOffset and are chained up until there are no more Table Entries left.
+ */
+@RequiredArgsConstructor
+class IndexReader {
+    //region Members
+
+    @NonNull
+    protected final ScheduledExecutorService executor;
+
+    //endregion
+
+    //region Index Access Operations
+
+    /**
+     * Gets the offset (from the given {@link SegmentProperties}'s Attributes up to which all Table Entries have been indexed.
+     *
+     * @param segmentInfo A {@link SegmentProperties} from which to extract the information.
+     * @return The offset.
+     */
+    long getLastIndexedOffset(SegmentProperties segmentInfo) {
+        return segmentInfo.getAttributes().getOrDefault(Attributes.TABLE_INDEX_OFFSET, 0L);
+    }
+
+    /**
+     * Locates the {@link TableBucket}s for the given Key Hashes in the given Segment's Extended Attribute Index.
+     *
+     * @param keyHashes A Collection of Key Hashes to look up {@link TableBucket}s for.
+     * @param segment   A {@link DirectSegmentAccess} providing access to the Segment to look into.
+     * @param timer     Timer for the operation.
+     * @return A CompletableFuture that, when completed, will contain the requested Bucket information.
+     */
+    CompletableFuture<Map<UUID, TableBucket>> locateBuckets(Collection<UUID> keyHashes, DirectSegmentAccess segment, TimeoutTimer timer) {
+        return segment
+                .getAttributes(keyHashes, false, timer.getRemaining())
+                .thenApply(attributes -> attributes.entrySet().stream()
+                        .collect(Collectors.toMap(Map.Entry::getKey, e -> new TableBucket(e.getKey(), e.getValue()))));
+    }
+
+    /**
+     * Looks up a Backpointer offset.
+     *
+     * @param offset  The offset to find a backpointer from.
+     * @param segment A DirectSegmentAccess providing access to the Segment to search in.
+     * @param timeout Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will contain the backpointer offset, or -1 if no such pointer exists.
+     */
+    CompletableFuture<Long> getBackpointerOffset(long offset, DirectSegmentAccess segment, Duration timeout) {
```

Comment:
This will be used in part 2 of this PR (the front-end), when `get` is invoked one a Key. This will be invoked if the sought key does not match the key to where its index entry points (i.e., we have a collision). If that's the case, this will traverse the list of backpointers until it finds something or there is nothing else to look up.

The chance of key hash collision is extremely small. In my tests, using SHA256 written into a 16-byte array generated no collisions for 100 million randomly generated keys.

Each call to this will be a Segment Attribute lookup. Segment Attributes have 2 layers of caching before a Tier 2 read is made:
1. In the Segment's Metadata - this is where all the changes are made, and attributes are kept until they are evicted (for lack of use).
2. In the Cache. The BTreeIndex caches pages in the RocksDB cache so for a warmed up index most of these lookups will not hit Tier 2.

When you'll see the second part of this PR (the front end), you'll see that I add a temporary (small) cache of backpointers too, which will take the brunt of all such lookups, if any.